### ISSUE TITLE
[no-use-before-define] Not considering hoisting

### ISSUE THREAD

**Issue by Tarnadas** [typescript-eslint/typescript-eslint#221]
The no-use-before-define rule is not considering hoisting, therefore giving me errors, when declaring a function below its usage.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-use-before-define": "error"
  }
}
```

```TS
fn(); // [eslint] 'fn' was used before it was defined. [@typescript-eslint/no-use-before-define]

function fn() {}
```

**Expected Result**

No linting error.

**Actual Result**

Linting error: [eslint] 'fn' was used before it was defined. [@typescript-eslint/no-use-before-define]

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.2.0` |
| `@typescript-eslint/parser`        | `1.2.0` |
| `TypeScript`                       | `3.2.4` |
| `ESLint`                           | `5.13.0` |
| `node`                             | `8.15.0` |

**Comment by platinumazure** [comment]
Hi @Tarnadas, thanks for the issue.

I can't speak to the typescript-eslint version of the rule, but in core ESLint, this would be functioning as designed. The entire point of this rule is to flag all uses before definitions-- *not* to flag if a variable use would be undefined. It is more of a stylistic/readability rule. The `no-undef` rule (or typescript-eslint's equivalent) detects if the value is actually not defined (for example, if you used `let fn = function() {}`, which is not hoisted.

Maintainers-- please correct me if I've misstated anything.

**Comment by Tarnadas** [comment]
Hey @platinumazure,
I just read through the ESLint page for this rule and it looks like you're right and this is WAI.

**Comment by aMarCruz** [comment]
@Tarnadas , use this:

```json
{
  "rules": {
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": ["error", {"functions": false, "typedefs": false}]
  }
```

Note there's [a bug](https://github.com/typescript-eslint/typescript-eslint/issues/249) with `typeof` ignoring `typedefs:false`, but once fixed the rule must work like intended.
